lang/src/org/partiql/lang/syntax/SqlParser.kt

```diff
@@ -912,11 +919,130 @@ class SqlParser(
                         if (isCrossJoin) metas + metaContainerOf(IsImplictJoinMeta.instance) else metas
                     )
                 }
+                ParseType.MATCH -> toGraphMatch()
                 else -> unwrapAliasesAndUnpivot()
             }
         }
     }
 
+    private fun ParseNode.toGraphMatch(): PartiqlAst.FromSource {
+        val metas = getMetas()
+
+        return PartiqlAst.build {
+            val expr = children[0].toAstExpr()
+            val patterns = children.tail.map {
+                if (it.type != ParseType.MATCH_EXPR) error("Invalid parse tree: expecting match expression in MATCH")
+                it.toGraphMatchPattern()
+            }
+
+            val matchExpr = PartiqlAst.GraphMatchExpr(patterns, metas = metas)
+            PartiqlAst.FromSource.GraphMatch(expr, matchExpr, metas)
+        }
+    }
+
+    private fun ParseNode.toGraphMatchPattern(): PartiqlAst.GraphMatchPattern {
+        val metas = getMetas()
+
+        return PartiqlAst.build {
+            val parts = children.map {
+                when (it.type) {
+                    ParseType.MATCH_EXPR_NODE -> it.toGraphMatchNode()
+                    ParseType.MATCH_EXPR_EDGE -> it.toGraphMatchEdge()
+                    else -> {
+                        TODO("Handle pattern part other than node&edge")
+                    }
+                }
+            }
+
+            // TODO quantifier
+            PartiqlAst.GraphMatchPattern(quantifier = null, parts = parts, metas = metas)
+        }
+    }
+
+    private fun ParseNode.toGraphMatchNode(): PartiqlAst.GraphMatchPatternPart.Node {
+        val metas = getMetas()
+
+        var name: SymbolPrimitive? = null
+        var label = mutableListOf<SymbolPrimitive>()
+        var predicate = null
```

Comment:
variables could be made immutable

```suggestion
        val label = mutableListOf<SymbolPrimitive>()
        val predicate = null
```